rust/src/storage/storage_with_flush_buffer.rs

```diff
@@ -0,0 +1,568 @@
+                        storage.set(id, &node)?;
+                    }
+                    Op::Delete => {
+                        // Delete the id in the underlying storage first, then remove it from the
+                        // flush buffer.
+                        // Once the id was deleted in the underlying storage layer, it may get
+                        // reused in a call to `reserve`.
+                        // For the case that the id was deleted in the storage layer and reassigned
+                        // by a concurrent task before the flush worker deleted it from the
+                        // flush buffer, `StorageWithFlushBuffer::reserve` also deletes the id
+                        // from the flush buffer. This way it is guaranteed, that a reassigned id
+                        // is not found in the flush buffer.
+                        storage.delete(id)?;
+                    }
+                }
+                flush_buffer.remove(&id);
+            } else {
+                // the buffer is currently empty
+                if shutdown.load(Ordering::SeqCst) {
+                    return Ok(());
+                }
+                // avoid busy looping
+                // TODO: use a condvar or similar
+                std::thread::yield_now();
+            }
+        }
+    }
+
+    pub fn shutdown(self) -> Result<(), Error> {
+        self.shutdown.store(true, Ordering::SeqCst);
+        for worker in self.workers {
+            worker.join().unwrap()?;
+        }
+        Ok(())
+    }
+}
+
+type FlushBuffer = DashMap<NodeId, Op>;
+
+/// An element in the flush buffer that can either be a set operation or a delete operation.
+#[derive(Debug, Clone)]
+enum Op {
+    Set(Node),
+    Delete,
+}
+
+/// A zero-sized type used as a certificate that a node was written to disk and can therefore be
+/// treated as clean again.
+// Note: The struct has to have a private field so that it cannot be instantiated outside of this
+// module.
+pub struct WriteCertificate(());
+
+#[cfg(test)]
+mod tests {
+    use std::{fs::File, sync::atomic::AtomicUsize, time::Duration};
+
+    use super::*;
+    use crate::{
+        storage::file::{
+            FileStorageManager, MockFileBackend, MockFileStorageManager, NoSeekFile,
+            NodeFileStorage,
+        },
+        types::NodeType,
+    };
+
+    #[test]
+    fn open_all_nested_layers() {
+        // The purpose of this test is to ensure that `StorageWithFlushBuffer` can be used with
+        // the lower layers of the storage system (that the types and interfaces line up).
+        let dir = tempfile::tempdir().unwrap();
+
+        // this opens:
+        // StorageWithFlushBuffer
+        //   -> FileStoreManager
+        //     -> NodeFileStorage for InnerNode
+        //       -> NoSeekFile
+        //     -> NodeFileStorage for LeafNode2
+        //       -> NoSeekFile
+        //     -> NodeFileStorage for LeafNode256
+        //       -> NoSeekFile
```

Comment:
```suggestion
        //     -> A NodeFileStorage for each node type (InnerNode, SparseLeafNode<N>, ...)
        //       -> NoSeekFile
```

The different node types aren't set in stone, so we shouldn't be too specific here.